source/extensions/filters/network/redis_proxy/command_splitter_impl.cc

```diff
@@ -900,6 +901,293 @@ void MSETRequest::onChildResponse(Common::Redis::RespValuePtr&& value, uint32_t
+          } else {
+            // If the current argument requires a value, add it to the request array
+            std::string value = incoming_request->asArray()[++i].asString();
+            addBulkString(requestArray, value);
+          }
+      }
+  }
+
+  // If the "count" argument is not found, add it with the default value
+  if (incoming_request->asArray().size() == 2) {
+      addBulkString(requestArray, "count");
+      addBulkString(requestArray, count);
+  }
+
+  std::unique_ptr<ScanRequest> request_ptr{
+      new ScanRequest(callbacks, command_stats, time_source, delay_command_latency)};
+
+  // caching the request and route for making child request from response
+  request_ptr->request_ = requestArray;
+  request_ptr->route_ = router.upstreamPool(key, stream_info);
+
+  if (request_ptr->route_) {
+    Extensions::NetworkFilters::RedisProxy::ConnPool::InstanceImpl* instance =
+        dynamic_cast<Extensions::NetworkFilters::RedisProxy::ConnPool::InstanceImpl*>(
+            request_ptr->route_->upstream(key).get());
+
+    request_ptr->num_of_Shards_ = instance->getNumofRedisShards();
+    if (request_ptr->num_of_Shards_ == 0 ) {
+      callbacks.onResponse(Common::Redis::Utility::makeError(Response::get().NoUpstreamHost));
+    }
+  }
+  else{
+    callbacks.onResponse(Common::Redis::Utility::makeError(Response::get().NoUpstreamHost));
+  }
+
+  // If shard index is some random value, we are setting the shard to 0 to avoid crashing
+  // This ensures scan always returns value.
+  if (shard_idx > request_ptr->num_of_Shards_) {
+    shard_idx = 0;
+  } else {
+    numofRequests = request_ptr->num_of_Shards_ - shard_idx + 1; 
+  }
+
+  // Reserving memory for pending_requests and pending_responses
+  // By default we are setting it to 3.
+  request_ptr->pending_requests_.reserve(numofRequests);
+  request_ptr->pending_responses_.reserve(numofRequests);
+
+  //keeping shard index and pending_requests_index same
+  for (int32_t i = request_ptr->num_of_Shards_-1; i >= shard_idx ; i--) {
+    request_ptr->pending_requests_.emplace_back(*request_ptr, i, i);
+  }
+
+  PendingRequest& pending_request = request_ptr->pending_requests_.back();
+  if (request_ptr->route_) {
+    pending_request.handle_= makeScanRequest(request_ptr->route_, shard_idx, requestArray, pending_request, callbacks.transaction());
+  }
+
+  if (!pending_request.handle_) {
+    pending_request.onResponse(Common::Redis::Utility::makeError(Response::get().NoUpstreamHost));
+    return nullptr;
+  }
+
+  return request_ptr;
+}
+
+void ScanRequest::onChildResponse(Common::Redis::RespValuePtr&& value, int32_t index, int32_t shard_index) {
+  // Request handled successfully
+  pending_requests_[index].handle_ = nullptr;
+  // Moving the response to pending response for doing validation later
+  // Incrementing the number of pending responses, it will drained during the validation
+  int64_t count = 2;
+
+  // Checking the cursor and number of objects for child request
```

Comment:
If response is Error , clear requests and responses if any and send response back immediately